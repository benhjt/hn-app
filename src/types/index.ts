export type StoryType = {
  id: number;
  title: string;
  points: number;
  content: string;
  user: string;
  time: number;
  time_ago: string;
  type: string;
  url: string;
  domain: string;
  comments_count: number;
  comments?: Array<CommentType>;
};

export type CommentType = {
  id: number;
  level: number;
  user: string;
  time: number;
  time_ago: string;
  content: string;
  comments?: Array<CommentType>;
};

export type UserType = {
  id: string;
  created: number;
  karma: number;
};
