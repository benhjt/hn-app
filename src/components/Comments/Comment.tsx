import { useEffect, useState } from 'react';
import { CommentType } from '../../types';
import { Link } from 'react-router-dom';
import Comments from './Comments';
import Separator from '../Separator';

const COLLAPSE_BUTTON_TEXT = '[-]';

const Comment = ({ comment }: { comment: CommentType }) => {
  const [collapsed, setCollapsed] = useState(false);
  const [collapseButton, setCollapseButton] = useState(COLLAPSE_BUTTON_TEXT);
  const [collapseTitle, setCollapseTitle] = useState('Hide replies');
  const [commentThreadSize, setCommentThreadSize] = useState(0);

  const toggleCollapse = () => {
    const newVal = !collapsed;
    setCollapsed(newVal);
    setCollapseButton(
      newVal ? `[${commentThreadSize} more]` : COLLAPSE_BUTTON_TEXT,
    );
    setCollapseTitle(
      `${newVal ? 'Show' : 'Hide'} ${commentThreadSize} comments`,
    );
  };

  useEffect(() => {
    const _calculateThreadSize = (thread: CommentType) => {
      let threadSize = 0;

      let flat = (thread: CommentType) => {
        threadSize++;
        thread?.comments?.forEach(flat);
      };

      flat(thread);
      return threadSize;
    };

    setCommentThreadSize(_calculateThreadSize(comment));
  }, [comment]);

  return (
    <div className="pl-5">
      <div className="flex py-1 text-gray-500 text-xs">
        <Link to={`/user/${comment.user}`} className="hover:underline">
          {comment.user}
        </Link>
        <span className="ml-2">{comment.time_ago}</span>
        <Separator />
        <button
          className="bg-transparent border-none cursor-pointer p-0 select-none"
          onClick={toggleCollapse}
          aria-label={collapseTitle}
        >
          {collapseButton}
        </button>
      </div>
      {collapsed ? (
        ''
      ) : (
        <>
          <div dangerouslySetInnerHTML={{ __html: comment.content }}></div>
          <Comments comments={comment.comments} />
        </>
      )}
    </div>
  );
};

export default Comment;
