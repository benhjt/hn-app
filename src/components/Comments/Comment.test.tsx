import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

import Comment from './Comment';
import { CommentType } from '../../types';

const comment: CommentType = {
  id: 37898999,
  level: 1,
  user: 'ingve',
  time: 1697459894,
  time_ago: '4 hours ago',
  content: 'hello',
};

describe('<Comment />>', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Router>
        <Comment comment={comment} />
      </Router>,
    );
    expect(container).toMatchSnapshot();
  });
});
