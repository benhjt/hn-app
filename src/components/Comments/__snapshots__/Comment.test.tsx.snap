// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comment />> renders correctly 1`] = `
<div>
  <div
    class="pl-5"
  >
    <div
      class="flex py-1 text-gray-500 text-xs"
    >
      <a
        class="hover:underline"
        href="/user/ingve"
      >
        ingve
      </a>
      <span
        class="ml-2"
      >
        4 hours ago
      </span>
      <div
        class="px-2"
      >
        |
      </div>
      <button
        aria-label="Hide replies"
        class="bg-transparent border-none cursor-pointer p-0 select-none"
      >
        [-]
      </button>
    </div>
    <div>
      hello
    </div>
    <div />
  </div>
</div>
`;
